cluster-autoscaler/processors/csi/csi_processor.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	apiv1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/autoscaler/cluster-autoscaler/context"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/kubernetes"
+	"k8s.io/client-go/informers"
+	storagev1beta1listers "k8s.io/client-go/listers/storage/v1beta1"
+	"k8s.io/klog/v2"
+)
+
+// CSIProcessor checks whether a node is ready for applications using volumes
+// provided by a CSI driver. This is relevant when the autoscaler has been
+// configured to check storage capacity.
+//
+// Without this processor, the following happens:
+// - autoscaler determines that it needs a new node to get volumes
+//   for a pending pod created
+// - the new node starts and is ready to run pods, but the CSI driver
+//   itself hasn't started running on it yet
+// - autoscaler checks for pending pods, finds that the pod still
+//   cannot run and asks for another node
+// - the CSI driver starts, creates volumes and the pod runs
+// => the extra node is redundant
+//
+// To determine whether a node will have a CSI driver, a heuristic is used: if
+// a template node derived from the node has a CSIStorageCapacity object, then
+// the node itself should also have one, otherwise it is not ready.
+type CSIProcessor interface {
+	// FilterOutNodesWithUnreadyResources removes nodes that should have a CSI
+	// driver, but don't have CSIStorageCapacity information yet.
+	FilterOutNodesWithUnreadyResources(context *context.AutoscalingContext, allNodes, readyNodes []*apiv1.Node) ([]*apiv1.Node, []*apiv1.Node)
+
+	// CleanUp frees resources.
+	CleanUp()
+}
```

Comment:
In contrast to `CustomResourceProcessor` there's nothing specific to CSI in this interface. A better solution might be to introduce a generic `FilterReadyNodes` interface and then support more than one of those in `AutoscalingProcessors`.

`CustomResourcesProcessor` then becomes an extension of that new interface which just adds `GetNodeResourceTargets`.